### PR TITLE
Added delay support in MailetContext.sendMail

### PR DESCRIPTION
**Pull request by apptaro**
I wonder if this is a good way but I created a pull request, so I can get some advice.

I sometimes want to postpone mail processing, such as when the database is temporarily down, I want to postpone the mail instead of raising an error.

Looking at the source code, I found that MailQueue supports delayed dequeue. So I made some changes so that MailetContext.sendMail can take delay parameters. Then custom mailets can utilize this functionality to postpone the mail processing.

### CODE REVIEW

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java [hunk 1]
```diff
@@ -404,16 +405,28 @@ public void sendMail(MailAddress sender, Collection<MailAddress> recipients, Mim
...
     @Override
     public void sendMail(Mail mail, String state) throws MessagingException {
         mail.setState(state);
         sentMails.add(fromMail(mail));
     }
 
+    @Override
+    public void sendMail(Mail mail, String state, long delay, TimeUnit unit) throws MessagingException {
+        mail.setState(state);
+        sentMails.add(fromMail(mail)); // FIXME delay ignored here for now
```
Comment (chibenwa) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: I think it is OK for now, this is a testing class. Maybe you can add delay + unit as a SentMail property?

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java [hunk 2]
```diff
@@ -27,6 +27,7 @@
 import java.util.Locale;
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java]: Could we add tests for these new sendMail methods?

**Comment by chibenwa** [comment]
It is more "Schedule delayed re-processing". This would nonetheless add a very interesting feature to mailet writers.

Could you create a JIRA for this?

Thanks,

Best regards,

Benoit

**Comment by apptaro** [comment]
Created the JIRA: https://issues.apache.org/jira/browse/JAMES-2271
Added a unit test to this pull request.

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/impl/JamesMailetContextTest.java [hunk 3]
```diff
@@ -214,6 +216,45 @@ public void sendMailShouldEnqueueEmailWithOtherStateWhenSpecified() throws Excep
...
+        mail.setMessage(MimeMessageBuilder.defaultMimeMessage());
+        testee.sendMail(mail, 5, TimeUnit.MINUTES);
+
+        ArgumentCaptor<Mail> mailArgumentCaptor = ArgumentCaptor.forClass(Mail.class);
+        ArgumentCaptor<Long> delayArgumentCaptor = ArgumentCaptor.forClass(Long.class);
+        ArgumentCaptor<TimeUnit> timeUnitArgumentCaptor = ArgumentCaptor.forClass(TimeUnit.class);
+        verify(spoolMailQueue).enQueue(mailArgumentCaptor.capture(), delayArgumentCaptor.capture(), timeUnitArgumentCaptor.capture());
+        verifyNoMoreInteractions(spoolMailQueue);
+
+        assertThat(mailArgumentCaptor.getValue().getState()).isEqualTo(Mail.DEFAULT);
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/impl/JamesMailetContextTest.java]: If you need several assertions, you should use `SoftAssertions`

1. Create the relevant junit class:

```
    @Rule
    public JUnitSoftAssertions softly = new JUnitSoftAssertions();
```

2. And then in the test with several assertThat: (those with a simple assertThat don't need it)

```
softly.assertThat(...)
softly.assertThat(...)
softly.assertThat(...)
....
```

It allows the display of all test error in a single run and makes debugging easier.

**Comment by chibenwa** [comment]
Nice, thanks ;-)

I have still a little remark...

**Comment by apptaro** [comment]
Thank you. I made changes to the unit test using JUnitSoftAssertions.

**Comment by chibenwa** [comment]
Hi!

This contribution had been merged. Thanks again for your support to the James project, your help is both useful and appreciable. Might you encounter any additional issues or want new features please feel free to contribute it.

As always, wouldn't you mind closing this pull request?

Best regards,

Benoit

**Pull request closed without merging**